File: Sean的leetcode/Leetcode/Maximum Product of Word Lengths.cpp
//Time:O(n^2)
class Solution {
public:
    int maxProduct(vector<string>& words) {
        unordered_map<int, int> table;
        int n = words.size();
        if (n == 0) return 0;
        int re = 0;
        for (auto s : words) {
            int mask = 0;
            for (auto c : s) {
                mask |= 1 << (c - 'a');
            }
            table[mask] = max(table[mask], (int)s.size());
            for (auto entry : table) {
                if (! (mask & entry.first)) {
                    re = max(re, entry.second * (int)s.size());
                }
            }
        }
        return re;
    }
}; 

File: Sean的leetcode/python/Maximum Product of Word Lengths.py
###
# Hash Table
# Time Complexity: O(n^2)
# Space Complexity: O(n)
###
class Solution(object):
    def maxProduct(self, words):
        """
        :type words: List[str]
        :rtype: int
        """
        if not words:
            return 0
        d = collections.defaultdict(int)
        maxlen = 0
        for word in words:
            key = 0
            for c in word:
                key |= 1 << (ord(c) - ord('a'))
            for k, l in d.iteritems():
                if key & k == 0:
                    maxlen = max(maxlen, l * len(word))
            d[key] = max(d[key], len(word))
        return maxlen
            
               
                    